### PR TITLE
Define `TypeParameter` and `ClassOrModule` classes; delete `Symbol` class.

### PR DESCRIPTION
**Pull request by jvilk-stripe**
Define `TypeParameter` and `ClassOrModule` classes; delete `Symbol` class.

### Motivation

Remove the monolithic `Symbol` class and instead have smaller symbol-type-specific classes with fewer ENFORCEs/weird corner cases/etc.

### Test plan

See included automated tests.

### CODE REVIEW

File: core/Symbols.h [hunk 1]
```diff
@@ -217,56 +217,117 @@ class Field final {
...
+
+    Flags flags;
+    // Method for TypeArgument, ClassOrModule for TypeMember.
+    SymbolRef owner;
+    NameRef name;
+    TypePtr resultType;
+
+private:
+    InlinedVector<Loc, 2> locs_;
+};
```
Comment (jvilk-stripe) [inline core/Symbols.h:297]: Before merging: Put in a `CheckSize` annotation.

File: core/Symbols.h [hunk 2]
```diff
@@ -217,56 +217,117 @@ class Field final {
...
+
+class ClassOrModule final {
```
Comment (jvilk-stripe) [inline core/Symbols.h:299]: Historically `Symbol` couldn't have `CheckSize` because `UnorderedMap` has different sizes in different environments. I dunno if that is still true.

Comment (froydnj) [reply core/Symbols.h:299]: The assertions from `CheckSize` are only active for `!debug_mode && !emscripten_build`...so I guess there'd have to be significant differences between x86-64 Linux and Mac (which doesn't seem likely)  to cause problems?  Maybe there would be latent issues for aarch64?

**Review by elliottt — APPROVED** [review]
This looks great!

**Comment by froydnj** [comment]
I don't understand why some of the tests around ambiguous definitions are failing.  My suspicion is that they are somehow related to the `isUndeclared` flag that got added prior to this PR being written originally, but I don't see where things would have gone wrong in the rebasing.  (Unless `isRewriterSynthesized` was also secretly being used for something related to classes/modules and it got unexpectedly nuked?)

**Review by elliottt — APPROVED** [review]
This looks great!

**Comment by froydnj** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_LUhyvVUv1zsJmu
→ https://go/builds/bui_LUhykf0KBMgmn5
